### PR TITLE
Enable cross-platform package signature verification for restore

### LINKED ISSUE #9004: XPlat Signing: Enable signature verification for restore in dotnet.exe
This is a tracking epic for getting verification working during restore. It involves the following:

1. Enable all verification-related tests, enabling signing tests as necessary to get existing verification working.
2. Make sure `nuget.exe verify` works on all platforms through mono
3. Make sure client policies work as expected on all platforms.

### PR DESCRIPTION
**Pull request by heng-liu**
## Bug

Fixes: https://github.com/NuGet/Home/issues/9004
Regression: No  
* Last working version:   
* How are we preventing it in future:   

## Fix

Details: 
1.Target NuGet.Packaging and projects depend on it to net5.0
2.Enable/implement cross-platform(net5.0 code path) package signing APIs (used in verification tests)
3.Enable/Implement cross-platform(net5.0 code path) package verification APIs.
4.Enable/fix related tests.

## Testing/Validation

Tests Added: Yes  
Reason for not adding tests:  
Validation:

### CODE REVIEW

File: build/common.targets
Comment (zivkan) [inline build/common.targets:20]: Can be done in another PR, but we should avoid nous with negative names, particularly when they represent a boolean type, as the double negative when the value is false can be confusing.

Also comparing against a list of TFMs is hard to read and maintain. We can use a condition something like ` '$(TargetFrameworkIdentifier)' == '.NETFramework' or ( '$(TargetFrameworkIdentifier)' == '.NETCoreApp' and [MSBuild]::VersionGreaterThanOrEquals('$(TargetFrameworkVersion)', '5.0') )`

Finally, this `SigningNotSupported` property appears to be used only to change the `DefineConstants` property below. Just use the condition on `DefineConstant`'s `PropertyGroup`. We don't need `SigningNotSupported`.

File: scripts/funcTests/runFuncTests.sh
Comment (zivkan) [inline scripts/funcTests/runFuncTests.sh]: delete dead/commented out code.

File: scripts/funcTests/runFuncTests.sh
Comment (zivkan) [inline scripts/funcTests/runFuncTests.sh:47]: Is it running twice because line 39 install the .NET Core 2.2 SDK?  These two executions of dotnet don't use the .NET 2.2 SDK installed into the cli folder, it's using whatever dotnet is already on the system path.

Either we don't need the .NET 2.2 SDK install on line 39, or we should change it to install a supported SDK version (.NET Core 2.2 went out of support quite some time ago), and then use the dotnet that was installed rather than the one on the path.

Also, .NET Core 3.0 and above no longer create a fallback folder, so when running those versions of the SDK, it should no longer be necessary to run twice to avoid the decompress and expand output being captured.

File: scripts/funcTests/runFuncTests.sh
Comment (zivkan) [inline scripts/funcTests/runFuncTests.sh]: This is unnecessary as long as line 39 installs the .NET Core 2.2 SDK.

File: scripts/funcTests/runFuncTests.sh
Comment (zivkan) [inline scripts/funcTests/runFuncTests.sh]: here we're using `$DOTNET`, but elsewhere, particularly lines 110 onwards, just uses `dotnet`. This inconsistency feels wrong.  I don't know how to reason about which version is used when.

File: scripts/utils/InstallCLIforBuild.ps1
Comment (zivkan) [inline scripts/utils/InstallCLIforBuild.ps1:23]: nitpick: If the install directory is passed in as a script parameter, and the build YAML passed in the env variable, then the script becomes more runabbled/debuggable locally.

File: scripts/utils/InstallCLIforBuild.sh
Comment (zivkan) [inline scripts/utils/InstallCLIforBuild.sh:5]: This would be better as a step in the yaml job, as spamming all env variables when trying to debug this script locally isn't helpful.

File: scripts/utils/InstallCLIforBuild.sh
Comment (zivkan) [inline scripts/utils/InstallCLIforBuild.sh]: delete dead/commented out code.

File: src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.Sdk2x.targets
Comment (zivkan) [inline src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.Sdk2x.targets]: This file's not used, right? We can delete it?

File: src/NuGet.Core/NuGet.Packaging/Signing/Timestamp/IRfc3161TimestampTokenInfo.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.Packaging/Signing/Timestamp/IRfc3161TimestampTokenInfo.cs:11]: very minor considering the type is `internal`, but why not put the `#if` outside the interface definition, so the interface only exists on TFMs that support signing?

File: src/NuGet.Core/NuGet.Packaging/Signing/Timestamp/Rfc3161TimestampProvider.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.Packaging/Signing/Timestamp/Rfc3161TimestampProvider.cs]: This is a breaking API change.  If we're going to break the API, can we make it `internal`, rather than `public`?  It doesn't seem like an API that people using `NuGet.Packaging` in their own apps would care about.

File: src/NuGet.Core/NuGet.Packaging/Signing/Timestamp/Rfc3161TimestampRequestNetstandard21Wrapper.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.Packaging/Signing/Timestamp/Rfc3161TimestampRequestNetstandard21Wrapper.cs:9]: The class defined in this function is wrapped in the same `#if` `#endif` block. Why are only these using statements excluded when the entire class hidden?  Just have a single `#if` on line 3/4.

File: test/NuGet.Core.FuncTests/Dotnet.Integration.Test/Dotnet.Integration.Test.csproj
Comment (zivkan) [inline test/NuGet.Core.FuncTests/Dotnet.Integration.Test/Dotnet.Integration.Test.csproj:14]: nitpick: trailing whitespace goes against coding guideline 23, and our `.editorconfig` file also instructs compatible editors to remove trailing whitespace, meaning someone is using an editor that does not respect `.editorconfig`.

File: build/common.project.props
Comment (zivkan) [inline build/common.project.props:23]: I think I've talked to you about when I worked on the dotnet.integration.tests multi-tfm testing: I'm not a fan of defining `RequiresSigningXplatAPIs` in the project, importing this props file, then using `TargetFrameworksLibrary` in the `csproj`'s `TargetFrameworks` element.

In my opnion, either:
* set properties in the project, import this file, and this file sets the `TargetFrameworks`

or

* import this file, which defines a bunch of different properties, and the project file uses the one it likes in its own definition of `TargetFrameworks`.

I prefer the second option, which is how `NuGet.Client` worked before the `dev-feature-xplatVerification` branch, but this current approach where some of the TFM logic is in the project file and other TFM logic is in the common project props file, it makes it harder to understand.

However, it's something we can fix after this branch is merged into `dev`.

**Review by zivkan — COMMENTED** [review]
There's a breaking API change that concerns me, otherwise it's just little things which could be improved in other PRs.

File: scripts/funcTests/runFuncTests.sh
Comment (heng-liu) [reply scripts/funcTests/runFuncTests.sh]: I think this is reasonable.
The dotnet version installed in runFuncTests are for testing. It's usually the latest version of the specified channel. Since it's installed with -NoPath, we can only use $DOTNET to refer to it. 
If using `dotnet`, it's referring to the one for build. It's installed in "Install .NET 5.0 for build" step in yaml file, usually the stable version of the specified channel. Since it's added to PATH. We can use `dotnet `to refer to it.
So:
```
	# Display current version
	$DOTNET --version
```
This is displaying the dotnet we installed, so use $DOTNET to see the version
```
# restore packages
dotnet msbuild build/build.proj /t:Restore
```
we would like to use the stable version to run restore. (we used to have problem when using the latest version to restore, but not sure if we still have problem for now)
```
# Unit tests
dotnet msbuild build/build.proj /t:CoreUnitTests
```
```
# Func tests
dotnet msbuild build/build.proj /t:CoreFuncTests /p:VisualStudioVersion=16.0 /p:Configuration=Release /p:BuildNumber=1 /p:ReleaseLabel=beta
```
As the target CoreUnitTests and CoreFuncTests will first run build on test projects, and we would like to build them using stable version,so we use `dotnet`.
When the testing dlls are generated, running those dlls, the dotnet matters is [here](https://github.com/NuGet/NuGet.Client/blob/dev-feature-xplatVerification/build/build.proj#L389), and it's value is https://github.com/NuGet/NuGet.Client/blob/dev/build/common.project.props#L24-L25

It's complicated and not intuitive, I should add some comments on it, if you agree with the usage of dotnet and $DOTNET

File: src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.Sdk2x.targets
Comment (heng-liu) [reply src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.Sdk2x.targets]: Good catch! Deleted.

File: scripts/funcTests/runFuncTests.sh
Comment (zivkan) [reply scripts/funcTests/runFuncTests.sh]: Line 84 of `runFuncTests.sh` is restoring "bootstrap" files needed for build. It is not running tests. As you wrote, the CoreUnitTests and CoreFuncTests targets need to find and use the dotnet cli installed in the `cli` folder. But `runFuncTests.sh` never runs tests directly and it's less confusing for it to use the same dotnet cli as the rest of the script.

File: src/NuGet.Core/NuGet.Packaging/Signing/Timestamp/Rfc3161TimestampProvider.cs
Comment (heng-liu) [reply src/NuGet.Core/NuGet.Packaging/Signing/Timestamp/Rfc3161TimestampProvider.cs]: I'm afraid we have to break this API.
There is no sync method for net5.0 code path(https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.pkcs.rfc3161timestamprequest?view=net-5.0) to send the timestamp request and return the timestamp.
So I have to use HttpClient.PostAsync to get the timestamp. That's why it's async.
I changed it into internal. 
Thanks!

File: scripts/funcTests/runFuncTests.sh
Comment (heng-liu) [reply scripts/funcTests/runFuncTests.sh]: Thanks! Fixed.

File: scripts/funcTests/runFuncTests.sh
Comment (heng-liu) [reply scripts/funcTests/runFuncTests.sh]: Thanks! Removed.

File: scripts/funcTests/runFuncTests.sh
Comment (heng-liu) [reply scripts/funcTests/runFuncTests.sh:47]: Thanks! Removed.

File: scripts/funcTests/runFuncTests.sh
Comment (heng-liu) [reply scripts/funcTests/runFuncTests.sh]: You're right. Changed to dotnet when restore bootstrap.proj. Thanks!

File: scripts/utils/InstallCLIforBuild.sh
Comment (heng-liu) [reply scripts/utils/InstallCLIforBuild.sh]: Thanks. Deleted.

File: scripts/funcTests/runFuncTests.sh
Comment (heng-liu) [reply scripts/funcTests/runFuncTests.sh:47]: Just found an error after removing run it twice.
The [log](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=3821073&view=logs&j=bd7d45da-bd9f-59a9-6834-84d44942bc5e&t=8712b7fc-4928-5354-8d68-786dcbf6d19e) shows that running dotnet command will display the following info first:
```
Welcome to .NET 5.0!
 ---------------------
 SDK Version: 5.0.100-preview.3.20216.6 Telemetry
 ---------
 The .NET tools collect usage data in order to help us improve your experience. The data is anonymous. It is collected by Microsoft and shared with the community. You can opt-out of telemetry by setting the DOTNET_CLI_TELEMETRY_OPTOUT environment variable to '1' or 'true' using your favorite shell.
 
 Read more about .NET CLI Tools telemetry: https://aka.ms/dotnet-cli-telemetry ----------------
 Installed an ASP.NET Core HTTPS development certificate.
 To trust the certificate run 'dotnet dev-certs https --trust' (Windows and macOS only).
 Learn about HTTPS: https://aka.ms/dotnet-https
 ----------------
 Write your first app: https://aka.ms/dotnet-hello-world
 Find out what's new: https://aka.ms/dotnet-whats-new
 Explore documentation: https://aka.ms/dotnet-docs
 Report issues and find source on GitHub: https://github.com/dotnet/core
 Use 'dotnet --help' to see available commands or visit: https://aka.ms/dotnet-cli
```
So the target got the wrong channel and version:
```
Channel is: Welcome
Version is: to
```
However, I could not repro on my local machine nor my Linux VM..

Comment (zivkan) [reply scripts/funcTests/runFuncTests.sh:47]: Sorry, I gave bad advice. Ignore this request. THe first time use banner is still displayed, even when it doesn't expand the fallbackfolder.

File: build/build.proj
Comment (nkolev92) [inline build/build.proj]: This seems like an unnecessary change, can you please revert it.

File: build/build.proj
Comment (nkolev92) [inline build/build.proj]: The whole file seems to have unnecessary changes.

File: build/common.targets
Comment (nkolev92) [reply build/common.targets:20]: We should create a follow up to clean it up.

File: scripts/funcTests/runFuncTests.sh
Comment (heng-liu) [reply scripts/funcTests/runFuncTests.sh:47]: No worries :) Added it back.

File: build/config.props
Comment (nkolev92) [inline build/config.props]: Do we still need the locking here? 

This whole part is super confusing to me :D

File: global.json
Comment (nkolev92) [inline global.json]: Can we upgrade to something newer? 
Do we need tihs for local development too or only the CI?

File: src/NuGet.Core/NuGet.Commands/NuGet.Commands.csproj
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/NuGet.Commands.csproj:14]: Does IsCore not work here?

File: src/NuGet.Core/NuGet.Packaging/Signing/Timestamp/Rfc3161TimestampTokenInfoNetstandard21Wrapper.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Packaging/Signing/Timestamp/Rfc3161TimestampTokenInfoNetstandard21Wrapper.cs:11]: Do we need this? We are only using net5 aren't we?

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Packaging/Signing/Timestamp/Rfc3161TimestampTokenInfoNetstandard21Wrapper.cs:11]: Or maybe we do need and should name it so something better :) 

Versions change :) 
Future Pr would be ok.

**Review by nkolev92 — COMMENTED** [review]
A few nits. 

It looks great beyond that. 

A product of lots of work over the last year. 

Great job @heng-liu , @zkat & @kartheekp-ms. 

Some kudos go to @zivkan too I guess.

**Review by nkolev92 — APPROVED** [review]
I'd wait for @zivkan's sign off before merging (well that and a green build)

**Comment by heng-liu** [comment]
> A few nits.
> 
> It looks great beyond that.
> 
> A product of lots of work over the last year.
> 
> Great job @heng-liu , @zkat & @kartheekp-ms.
> 
> Some kudos go to @zivkan too I guess.

Yes, @zivkan helps a lot :)
And we also would like to thank for @dtivel
You and @rrelyea also helped a lot when targeting to netstandard2.1 ;)

File: test/NuGet.Core.Tests/NuGet.Frameworks.Test/NuGet.Frameworks.Test.csproj
Comment (nkolev92) [inline test/NuGet.Core.Tests/NuGet.Frameworks.Test/NuGet.Frameworks.Test.csproj:11]: I don't think we need to test for 2 different frameworks in a project like NuGet.Frameworks. 

There is no signing code in here. 

Same probably apllies to a bunch of the other projects.

File: test/NuGet.Core.Tests/NuGet.LibraryModel.Tests/NuGet.LibraryModel.Tests.csproj
Comment (nkolev92) [inline test/NuGet.Core.Tests/NuGet.LibraryModel.Tests/NuGet.LibraryModel.Tests.csproj:11]: Likely requiresSigningXplatAPIs is not needed here either.

File: build/build.proj
Comment (heng-liu) [reply build/build.proj]: Thanks! Fixed.

File: build/build.proj
Comment (heng-liu) [reply build/build.proj]: Thanks! Fixed.

File: build/config.props
Comment (heng-liu) [reply build/config.props]: You're right. We just use the latest version now. Removed the lock part.
Thanks!

File: global.json
Comment (heng-liu) [reply global.json]: Updated to public preview4.
When the bug in public preview5 got fix, we will update to that version.
Created a tracking issue https://github.com/NuGet/Client.Engineering/issues/367

**Review by zivkan — APPROVED** [review]
There are several things I would love to have cleaned up, but we can do that after this is merged in dev.

**Comment by heng-liu** [comment]
Thanks. 
Yes, there're some comments I haven't addressed. Will do that after this merged in dev.

File: src/NuGet.Core/NuGet.Commands/NuGet.Commands.csproj
Comment (heng-liu) [reply src/NuGet.Core/NuGet.Commands/NuGet.Commands.csproj:14]: `IsCore `doesn't work.
Properties are evaluated in the order in which they appear in the project file. 
The assignment of property `IsCore `is in common.targets.
And the NuGet.commands.csproj imports the common.targets at the end, that's after the `NoWarn`.

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/NuGet.Commands.csproj:14]: We can probably move IsCore to common.props...I can look into that clean-up.

File: test/NuGet.Core.FuncTests/Dotnet.Integration.Test/Dotnet.Integration.Test.csproj
Comment (heng-liu) [reply test/NuGet.Core.FuncTests/Dotnet.Integration.Test/Dotnet.Integration.Test.csproj:14]: Thanks. Will fix in https://github.com/NuGet/NuGet.Client/pull/3571

File: build/common.project.props
Comment (heng-liu) [reply build/common.project.props:23]: Thanks. Will fix in #3571

File: test/NuGet.Core.Tests/NuGet.LibraryModel.Tests/NuGet.LibraryModel.Tests.csproj
Comment (heng-liu) [reply test/NuGet.Core.Tests/NuGet.LibraryModel.Tests/NuGet.LibraryModel.Tests.csproj:11]: Thanks. Will fix in #3571

File: test/NuGet.Core.Tests/NuGet.Frameworks.Test/NuGet.Frameworks.Test.csproj
Comment (heng-liu) [reply test/NuGet.Core.Tests/NuGet.Frameworks.Test/NuGet.Frameworks.Test.csproj:11]: Thanks. Will fix in #3571

File: src/NuGet.Core/NuGet.Packaging/Signing/Timestamp/Rfc3161TimestampTokenInfoNetstandard21Wrapper.cs
Comment (heng-liu) [reply src/NuGet.Core/NuGet.Packaging/Signing/Timestamp/Rfc3161TimestampTokenInfoNetstandard21Wrapper.cs:11]: I think the logic of `#if IS_SIGNING_SUPPORTED && IS_CORECLR` is more clear than` #if NET5_0`.
We can tell the first one, is related to signing, and when it's the netcore code path.
The second one may be related to signing, or may be related to net5.0 API change unrelated to signing.

File: src/NuGet.Core/NuGet.Packaging/Signing/Timestamp/Rfc3161TimestampRequestNetstandard21Wrapper.cs
Comment (heng-liu) [reply src/NuGet.Core/NuGet.Packaging/Signing/Timestamp/Rfc3161TimestampRequestNetstandard21Wrapper.cs:9]: Thanks. Will fix in #3571

File: src/NuGet.Core/NuGet.Packaging/Signing/Timestamp/IRfc3161TimestampTokenInfo.cs
Comment (heng-liu) [reply src/NuGet.Core/NuGet.Packaging/Signing/Timestamp/IRfc3161TimestampTokenInfo.cs:11]: Thanks. Will fix in #3571

File: build/common.targets
Comment (heng-liu) [reply build/common.targets:20]: I tried the following, but the build failed, as `TargetFrameworkIdentifier `and `TargetFrameworkVersion` are empty.
```
<PropertyGroup Condition=" '$(IsDesktop)' == 'true' or ( '$(TargetFrameworkIdentifier)' == '.NETCoreApp' and $([MSBuild]::VersionGreaterThanOrEquals('$(TargetFrameworkVersion)', '5.0')) ) ">
<DefineConstants>$(DefineConstants);IS_SIGNING_SUPPORTED</DefineConstants>
</PropertyGroup>
```
I then tested two projects with `TargetFrameworks` and `TargetFramework`, looks like the later one got TargetFrameworkIdentifier  and TargetFrameworkVersion correctly, the former one didn't.

**Pull request merged**
